File: LLD/Generics/Generic_LinkedList/LinkedList.java
package LLD.Generics.Generic_LinkedList;
/*
 * Generic Singly Linked List Implementation
Problem Statement
Your task is to implement a generic singly linked list in Java, capable of storing data of any type. The class ListNode has been provided, which represents the nodes of the linked list. You are required to complete the implementation of the LinkedList class, which should include methods for adding elements to the list, displaying the list contents, checking if the list is empty, and determining the size of the list.

Instructions
Inside the LinkedList class, initialize the head node to null in the constructor. The head node will be used to keep track of the start of the list.
Implement the add method to add a new node with the specified data at the end of the list. This method should take the data to be added as a parameter and does not return anything.
Implement the display method to print the data of each node in the list, separated by spaces. Ensure that no additional spaces or newline characters are printed at the end of the output.
Implement the isEmpty method to return true if the list is empty (i.e., the head is null), otherwise false. This method does not take any parameters and returns a boolean value.
Implement the size method to return the number of nodes currently in the list. This method does not take any parameters and returns an integer value representing the size of the list.
 */
class ListNode<T> {
    T data;
    ListNode<T> next;

    public ListNode(T data) {
        this.data = data;
        this.next = null;
    }
}

public class LinkedList<T> {
    private ListNode<T> head;

    // Constructor to initialize head as null
    public LinkedList() {
        this.head = null;
    }

    // Method to add a new node at the end of the list
    public void add(T data) {
        ListNode<T> newNode = new ListNode<>(data);
        if (head == null) {
            head = newNode;
        } else {
            ListNode<T> temp = head;
            while (temp.next != null) {
                temp = temp.next;
            }
            temp.next = newNode;
        }
    }

    // Method to display the elements of the linked list
    public void display() {
        if (head == null) {
            System.out.print(""); // Ensure empty output for an empty list
            return;
        }

        StringBuilder result = new StringBuilder();
        ListNode<T> temp = head;

        while (temp != null) {
            result.append(temp.data).append(" "); // Append each element with a trailing space
            temp = temp.next;
        }

        System.out.print(result.toString()); // Output without an extra newline
    }

    // Method to check if the linked list is empty
    public boolean isEmpty() {
        return head == null;
    }

    // Method to return the number of nodes in the linked list
    public int size() {
        int count = 0;
        ListNode<T> temp = head;
        while (temp != null) {
            count++;
            temp = temp.next;
        }
        return count;
    }
}

File: LLD/Generics/Generic_LinkedList/linkedList.py
# Generic Linked List Implementation
# Problem Statement
# You are tasked with implementing a generic linked list data structure in Python. The linked list should be able to store elements of any type and provide methods for appending new elements and printing the list.

# Requirements
# The implementation should include the following components:

# Node class:
# A generic class that represents a node in the linked list.
# Each node should have a data attribute to store the element and a next attribute to store the reference to the next node.
# LinkedList class:
# A generic class that represents the linked list.
# Should have a head attribute to store the reference to the first node and a tail attribute to store the reference to the last node.
# Should have an append method to add a new element to the end of the list.
# Should have a print_list method to print the elements of the list in the order they are stored.
# The implementation should be type-safe and support elements of any type.
# Instructions
# Implement the Node class according to the specified requirements.
# Implement the LinkedList class with the required attributes and methods.
# Ensure that the implementation is type-safe by using type annotations and the typing module.

from typing import TypeVar, Generic, Optional

T = TypeVar('T')

class Node(Generic[T]):
    def __init__(self, data: T):
        self.data: T = data
        self.next: Optional[Node[T]] = None  # Reference to the next node

class LinkedList(Generic[T]):
    def __init__(self):
        self.head: Optional[Node[T]] = None  # First node reference
        self.tail: Optional[Node[T]] = None  # Last node reference

    def append(self, data: T):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
            self.tail = new_node
        else:
            assert self.tail is not None  # For type safety
            self.tail.next = new_node
            self.tail = new_node

    def print_list(self):
        current_node = self.head
        while current_node:
            print(current_node.data, end=" -> ")
            current_node = current_node.next
        print("None")

# Example usage:

# Create a linked list of integers
ll_int = LinkedList[int]()
ll_int.append(1)
ll_int.append(2)
ll_int.append(3)
ll_int.print_list()  # Output: 1 -> 2 -> 3 -> None

# Create a linked list of strings
ll_str = LinkedList[str]()
ll_str.append("a")
ll_str.append("b")
ll_str.append("c")
ll_str.print_list()  # Output: a -> b -> c -> None
